lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchResourceManager.java

```diff
@@ -77,22 +80,48 @@ public void onResourceRequested(final ResourceRequestEvent resourceRequestEvent)
   public void onResourceReleased(final ResourceReleaseEvent resourceReleaseEvent) {
     LOG.log(Level.FINEST, "Got ResourceReleasedEvent for Id: {0} in AzureBatchResourceManager",
         resourceReleaseEvent.getIdentifier());
-    this.containerRequests.remove(resourceReleaseEvent.getIdentifier());
+    if (!this.activeContainerIds.remove(resourceReleaseEvent.getIdentifier())) {
+      LOG.log(Level.WARNING,
+          "Attempting to remove non-existent activeContainer for Id: {0} in AzureBatchResourceManager",
+          resourceReleaseEvent.getIdentifier());
+    }
+
+    if (this.containerRequests.containsKey(resourceReleaseEvent.getIdentifier())) {
```

Comment:
This code is not thread safe. If two threads are performing the same action at the same time you'll get race conditions. If you really want to add logging, you can just call "remove" and check if the return value is null.

Reply:
Fixed.